tests/test_visual_typing_char_by_char.py

```
@@ -55,30 +55,52 @@ def main() -> int:
+            # The chain keystroke -> PTY echo -> Ghostty render -> committed frame -> snapshot
+            # diff is asynchronous; under CI/VM load a single committed frame can take well
+            # over 120ms. Instead of a fixed sleep + one hard assert, poll the real signals:
+            # the rendered pixel diff crossing min_pixels AND the terminal text buffer holding
+            # the typed prefix. Each panel_snapshot diffs against the prior call, so a frame can
+            # commit between two polls; latch on the first snapshot that crosses the threshold so
+            # a split diff across polls still counts. Fail only at the deadline.
+            expected_prefix = text[: i + 1]
+            state = {"changed": -1, "snap": None, "buf": ""}
+
+            def _typed_visible() -> bool:
+                snap = c.panel_snapshot(panel_id, f"typing_{i}_after_{ord(ch)}")
```

Ruff (0.15.17)

[warning] Function definition does not bind loop variable `i`

(B023)